v-next/core/test/internal/tasks/task-manager.ts

```diff
@@ -0,0 +1,908 @@
+      "Should have posParam1",
+    );
+    assert.ok(
+      task1.positionalParameters.some((p) => p.name === "posParam1"),
+      "Should have varParam1",
+    );
+    // New params should be added by the overrides
+    assert.ok(task1.namedParameters.get("param2"), "Should have param2");
+    assert.ok(task1.namedParameters.get("flag2"), "Should have flag2");
+    assert.ok(task1.namedParameters.get("param3"), "Should have param3");
+    assert.ok(task1.namedParameters.get("flag3"), "Should have flag3");
+    // Should have 3 actions
+    assert.equal(task1.actions.length, 3);
+    assert.equal(task1.actions[0].pluginId, "plugin1");
+    assert.equal(task1.actions[1].pluginId, "plugin1");
+    assert.equal(task1.actions[2].pluginId, "plugin2");
+  });
+
+  it("should add an empty task", async () => {
+    const hre = await createHre({
+      plugins: [
+        {
+          id: "plugin1",
+          tasks: [
+            {
+              id: ["task1"],
+              description: "description1",
+              type: TaskDefinitionType.EMPTY_TASK,
+            },
+          ],
+        },
+      ],
+    });
+
+    const task1 = hre.tasks.getTask("task1");
+    assert.deepEqual(task1.id, ["task1"]);
+    assert.equal(task1.pluginId, "plugin1");
+  });
+
+  it("should add subtasks", async () => {
+    const hre = await createHre({
+      plugins: [
+        {
+          id: "plugin1",
+          tasks: [
+            {
+              id: ["task1"],
+              description: "description1",
+              type: TaskDefinitionType.EMPTY_TASK,
+            },
+            // adds a subtask to the empty task
+            new NewTaskDefinitionBuilderImplementation(["task1", "subtask1"])
+              .setAction(() => {})
+              .build(),
+          ],
+        },
+
+        {
+          id: "plugin2",
+          tasks: [
+            // adds a subtask to the non-empty task
+            new NewTaskDefinitionBuilderImplementation([
+              "task1",
+              "subtask1",
+              "subsubtask1",
+            ])
+              .setAction(() => {})
+              .build(),
+          ],
+        },
+      ],
+    });
+
+    const task1 = hre.tasks.getTask("task1");
+    assert.deepEqual(task1.id, ["task1"]);
+    assert.equal(task1.pluginId, "plugin1");
+
+    const subtask1 = hre.tasks.getTask(["task1", "subtask1"]);
+    assert.deepEqual(subtask1.id, ["task1", "subtask1"]);
+    assert.equal(subtask1.pluginId, "plugin1");
+
+    const subsubtask1 = hre.tasks.getTask(["task1", "subtask1", "subsubtask1"]);
+    assert.deepEqual(subsubtask1.id, ["task1", "subtask1", "subsubtask1"]);
+    assert.equal(subsubtask1.pluginId, "plugin2");
+  });
+
+  /**
+   * These are all tested with plugin tasks, but the same logic applies to config tasks
+   */
+  describe("errors", () => {
+    it("should throw if there's a global parameter with the same name as a task named parameter", async () => {
```

Comment:
Also positional

Reply:
Currently, it's only checking against named params, see https://github.com/NomicFoundation/hardhat/blob/6413a98470ba3c8671b41f5d90342ada56ea6cce/v-next/core/src/internal/tasks/task-manager.ts#L188

Does it also need to validate against positional parameters? If so, I can add the validation & tests in a follow-up PR.

Reply:
Yes, i think it's better to be conservative. A follow up pr looks better.